### DISCUSSION TITLE
Are invisible imports are thing?

### DISCUSSION THREAD

**Discussion by krismc** [micropython/micropython#16485] [General]
Apologies if this question is so dumb it offends but we all have to start somewhere. To setup the ULP on the esp32s2 I have to do a once only
```
mip.install('github:micropython/micropython-esp32-ulp')
```
This makes an esp32_ulp folder on the esp32 & populates it with a bunch of files
```
>>> os.listdir('lib/esp32_ulp')
['__init__.py', '__int__.py', '__main__.py', '_main__.py', 'assemble.py', 'nocomment.py', 'opcodes.py', 'opcodes_s2.py', 'parse_to_db.py', 'preprocess.py', 'soc.py', 'soc_s2.py', 'soc_s3.py', 'util.py']
```
The install must do something else too because if try just copying those 14 files into the esp32 (in lieu of doing an install) the user ulp code crashes at
'''
from esp32_ulp import src_to_binary
'''

What I don't understand is even when I mip.install there is no src_to_binary in the esp32_ulp folder. How is it the user code can import something that isn't there?

**Comment by scruss** [comment]
> there is no src_to_binary in the esp32_ulp folder

Look in [__init__.py](https://github.com/micropython/micropython-esp32-ulp/blob/master/esp32_ulp/__init__.py#L26). It's there

**Reply by krismc** [reply]
Tnx, I thought function imports had to be something like  from __init__.py import src_to_binary so I was expecting an src_to_binary.py file

My experience with repositories is they eventually stop working. Is there a way I could setup that esp32_ulp folder from my PCs hard drive rather than have to get each new esp32 online to do the mip.install?
